framework/db/ActiveRecord.php

```diff
@@ -199,34 +200,56 @@ protected static function findByCondition($condition)
      * This method will ensure that an array condition only filters on existing table columns.
      *
      * @param array $condition condition to filter.
+     * @param array $aliases
      * @return array filtered condition.
      * @throws InvalidArgumentException in case array contains unsafe values.
+     * @throws InvalidConfigException
      * @since 2.0.15
      * @internal
      */
-    protected static function filterCondition(array $condition)
+    protected static function filterCondition(array $condition, array $aliases = [])
     {
         $result = [];
         $db = static::getDb();
-        // valid column names are table column names or column names prefixed with table name
+        $columnNames = static::filterValidColumnNames($db, $aliases);
+
+        foreach ($condition as $key => $value) {
+            if (is_string($key) && !in_array($db->quoteSql($key), $columnNames, true)) {
+                throw new InvalidArgumentException('Key "' . $key . '" is not a column name and can not be used as a filter');
+            }
+            $result[$key] = is_array($value) ? array_values($value) : $value;
+        }
+
+        return $result;
+    }
+
+    /**
+     * Valid column names are table column names or column names prefixed with table name or table alias
+     *
+     * @param Connection $db
+     * @param array $aliases
+     * @return array
+     * @throws InvalidConfigException
+     */
+    protected static function filterValidColumnNames($db, array $aliases)
+    {
         $columnNames = [];
         $tableName = static::tableName();
         $quotedTableName = $db->quoteTableName($tableName);
+        $tableAliases = array_keys(ArrayHelper::removeValue($aliases, $tableName, []));
```

Comment:
Why does it work in this way? Why not `array_unique(array_keys($aliases))`?

Reply:
If I remember well, then it is to prevent to check twice for the table name and for the alias, if these two are the same.

Reply:
I don't get it. AFAIK `$aliases` could be array like `['t1' => 'table1', 't2' => 'table2']`. Assuming that we're in `Table1` model, this will ignore `t2` alias - why?

Reply:
You are right, I messed this up. I've just refactored it, based on that aliases are also could be `['t1' => 'table1', 't2' => 'table1']`.

Also I skip the alias if it is the same as the table name eg.